New Tools: HTTP pre-built tools

Hormold opened the pull request:
I'd like to propose two new beta tools that address common patterns I've seen when building voice agents.

~~1. SMS Tool - Many voice agents need to send text messages during calls (order confirmations, addresses, booking links, etc.). This tool provides a simple way to send SMS via Twilio, SignalWire, or Vonage. It can auto-detect the caller's phone number from the SIP session or use pre-configured recipients. The goal is to reduce boilerplate for this common use case while keeping it flexible enough for different workflows.~~
UPD: **SMS tools moved to another branch to separate discussion**

HTTP Tool - A lot of agents need to call external APIs - webhooks, CRMs, internal services, etc. Rather than writing custom functions for each HTTP call, this tool lets you declaratively configure requests with parameters, headers, timeouts, and response handling. The idea is to make simple API integrations faster to implement while still supporting customization when needed.

Both tools include error handling, provider auto-detection (SMS), and parameter validation. I've included tests and an example with the weather agent showing how they work together.

Open to feedback on the API design, naming, and what additional features might be useful!

bcherry commented on livekit-agents/livekit/agents/beta/tools/http/send_http.py:
```diff
@@ -0,0 +1,306 @@
+    """Make HTTP request with given parameters.
+
+    Args:
+        url: Endpoint URL
```
in agent builder work I found that it's a must-have to support templating arguments into named path segments if this is going to work for any REST API. so this shouldn't be an actual URL but should be a URL template, and we'd need a way to partition the passed arguments into those that go into the URL and those that go into the payload

bcherry commented on livekit-agents/livekit/agents/beta/tools/http/send_http.py:
```diff
@@ -0,0 +1,306 @@
+        raise ValueError("Invalid parameters schema - must be a dict with 'type' field")
+
+    tool_name = config.name or "http_request"
+    tool_description = config.description or f"Make {config.method} request to {config.url}"
```
i don't think a default description makes sense - if the dev leaves it blank, it should remain blank

bcherry commented on livekit-agents/livekit/agents/beta/tools/http/send_http.py:
```diff
@@ -0,0 +1,306 @@
+                },
+                "required": ["city"]
+            },
+            output_normalizer=lambda resp: f"Temp: {resp.body}" if resp.success else None
```
this syntax seems pretty verbose - i wonder if there's a more pythonic way to do it?

bcherry commented on livekit-agents/livekit/agents/beta/tools/http/send_http.py:
```diff
@@ -0,0 +1,306 @@
+    method = method.upper()
+    timeout = aiohttp.ClientTimeout(total=timeout_ms / 1000.0)
+
+    async with aiohttp.ClientSession(timeout=timeout) as session:
```
this should use `utils.http_session`

bcherry commented on livekit-agents/livekit/agents/beta/tools/http/send_http.py:
```diff
@@ -0,0 +1,306 @@
+        timeout_ms: Request timeout in milliseconds (default: 30000)
+        parameters: JSON schema dict for tool parameters (defaults to empty object schema)
+        execution_message: Optional message to announce before executing (spoken by agent)
+        output_normalizer: Optional function to transform HTTPResponse to string or None
```
rather than adding additional complexity to the tool - it may be better to create two layers of abstraction:

1. a simplified utility for performing a URL request from a tool call that handles variable templating and errors consistently, and could be used inside any function_tool
2. a tool factory method that takes an HTTP config and outputs a simple tool using the aforementioned utility

this way devs could add any kind of before/after customization they need by leveraging the lower-level utility, but we'd still offer a no-frills basic wrapper

Hormold commented:
- added url templating with :param placeholders, auto-validating that every path param is declared and stripping those fields from the body. if the param is part of the URL template it's substituted into the path, otherwise it stays in the payload.
- built a quick schema builder + arg sanitizer so only whitelisted, type-checked params go over the wire; bad inputs from LLM raise ToolError.
- switched to the shared utils http session and block interruptions for mutating requests
- split the code into a reusable run_http_request() helper plus the existing factory, so custom tools can hook before/after without rewriting http logic.
- http tool name is required, no default description

so now you can either pass a full json_schema dict for complete control, or a shorthand params list (HTTPToolParam) for common string/enum/int/bool cases - the factory turns both into the same schema before validation.
